### PR TITLE
Kerb Validator 

### LINKED ISSUE #1460: Add validation to create kerb nodes for pedestrian crossings.
### Description

While mapping sidewalks and crossings, we should consider adding automagic validations for creating kerb nodes wherever crossing ways meeting sidewalks. 

We could also consider a power user feature that would just create the kerb nodes with extreme prejudice. According to chad and Said, there is basically never going to be a sidewalk / crossing way intersection without a kerb node needing to be there. 

Ex: 
This user is in the process of drawing a crossing way and has left it *close* to the sidewalk, and rapid has generated a lint showing the fact that they should probably be connected. 
image

After connecting the two features, we should at least put up a lint saying that there should probably be a kerb node here: 

image

### PR DESCRIPTION
**Pull request by tannerwuster**
- [x] Flag any crossings without kerb nodes with a new validator.
- [x] The validator will give the user a button to 'fix' by adding kerb nodes 1m from the endpoints of the crossing. (It will be up to the user to position the kerb nodes.)

- [x] Offer different fixes in the validation pane
- Flush
- Lowered
- Raised

Screenshot 2024-10-11 at 10 34 38

closes #1460

### CODE REVIEW

**Comment by eneerhut** [comment]
When warnings are opened up, do you think there could be value in numbering each warning and having a corresponding yellow circle with a number surrounding the feature? You currently have to click through to reveal which warning corresponds to the feature.

File: modules/validations/kerb_nodes.js [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+
+
+  /**
+   * hasRoutableTags
+   * Checks if the given way has tags that make it routable
+   * @param  {Object} way - The way entity to check
+   * @return {Boolean} True if the way has routable tags, false otherwise
+   */
+  function hasRoutableTags(way) {
+    const routableTags = ['highway', 'railway', 'waterway'];
```
Comment (bhousel) [inline modules/validations/kerb_nodes.js:71]: We probably only want to consider `highway` here.

File: modules/validations/kerb_nodes.js [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+   * intersectsPedestrianPathway
+   * Determines if the given way intersects with pedestrian pathways
+   * @param  {Object} way - The way entity to check
+   * @param  {Object} graph - The graph containing the way and node data
+   * @return {Boolean} True if the way intersects pedestrian pathways, false otherwise
+   */
+  function intersectsPedestrianPathway(way, graph) {
+    let intersectingWays = new Set();
+    way.nodes.forEach(nodeId => {
+      if (!nodeId) {
```
Comment (bhousel) [inline modules/validations/kerb_nodes.js:86]: Can this really happen?

File: modules/validations/kerb_nodes.js [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+   * Displays a reference for the issue in the UI
+   * @param  {Object} selection - The UI selection to append the reference to
+   */
+  function showReference(selection) {
+    selection.selectAll('.issue-reference')
+      .data([0])
+      .enter()
+      .append('div')
+      .attr('class', 'issue-reference')
+      .html(l10n.tHtml('issues.kerb_nodes.reference'));
```
Comment (bhousel) [inline modules/validations/kerb_nodes.js:128]: Always use `.text` and `.t` for these unless there's a really good reason to inject html into the document (doing this is kind of unsafe - someone could translate this string and insert malicious html).

```suggestion
      .text(l10n.t('issues.kerb_nodes.reference'));
```

**Review by bhousel — APPROVED** [review]
This all looks really awesome - I like that it also considers refuge situations where the sidewalk have different kinds of junctions and refuge islands.
And includes tests too!  I can't wait to try it out...

File: data/core.yaml [hunk 4]
```diff
@@ -1874,6 +1874,11 @@ en:
       tip: "Find way segments that are duplicated (same node endpoints)"
       message: "Two or more segments share the same pair of nodes"
       reference: "Redundant way segments should be merged or moved apart."
+    kerb_nodes:
```
[the remark this answers is not in the document]
Comment (tannerwuster) [reply data/core.yaml:1877]: I was also confused on why it was spelt 'kerb.' Updating this now!

**Comment by bhousel** [comment]
Looks good, lets merge it and then people can try it out in canary! 🐤

**Comment by eneerhut** [comment]
Nice work @tannerwuster! Looking great in Canary so far.
Could we add a link to the [kerb tagging guidelines](https://wiki.openstreetmap.org/wiki/Key:kerb#:~:text=is%20being%20crossed.-,Values,-%5Bedit%20%7C) in the OSM wiki when the user clicks the 'i'?
image

**Pull request merged**